File: 0125-valid-palindrome/0125-valid-palindrome.py
class Solution(object):
    def isPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        n = len(s)
        i = 0
        j = n-1
        while i<j:
            while i<j and not s[i].isalnum(): i+=1
            while i<j and not s[j].isalnum(): j-=1
            if s[i].lower() != s[j].lower(): return False # be case-insensitive
            i+=1
            j-=1
        return True
        

File: 0125-valid-palindrome/0125-valid-palindrome.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        int n = s.size();
        // two pointers
        int i = 0, j = n - 1;

        // invariant: all (alphanumeric) characters strictly outside of [i, j] have been checked and matched
        while (i < j){
            // skip over any non-alphanumeric characters
            while (i < j && !isalnum((unsigned char) s[i])) ++i; // char is signed on some platforms so cast to unsigned for safety
            while (i < j && !isalnum((unsigned char) s[j])) --j;

            if (tolower((unsigned char) s[i]) != tolower((unsigned char) s[j])) // be case-insensitive
                return false; 
            
            ++i; --j;           
        }

        return true;
    }
};
